### PR TITLE
Support individual MSP updates for orderer organization

### PR DESCRIPTION
**Pull request by stephyee**
#### Type of change

- New feature

#### Description

Support individual MSP updates for orderer organization

#### Related issues

[FAB-18063](https://jira.hyperledger.org/browse/FAB-18063)

### CODE REVIEW

File: configtx/application.go [hunk 1]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveRootCert removes an trusted root certificate from the organization MSP.
```
Comment (caod123) [inline configtx/application.go:501]: `removes a trusted`

File: configtx/application.go [hunk 2]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+
+	certs := msp.RootCerts[:]
+	for i, c := range msp.RootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.RootCerts = certs
```
Comment (caod123) [inline configtx/application.go:516]: You may want to validate the ca certs here as well since removal might break the intermediate cert chaining. Probably add a specific error indicating that removal of the root cert would invalidate an intermediate cert.

File: configtx/application.go [hunk 3]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveIntermediateCert removes an trusted intermediate certificate trusted by the organization MSP.
```
Comment (caod123) [inline configtx/application.go:544]: `removes a trusted`

File: configtx/application.go [hunk 4]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+			break
+		}
+	}
+
+	msp.IntermediateCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddOUIdentifier adds a fabric organizational unit for the organization MSP.
```
Comment (caod123) [inline configtx/application.go:564]: `AddOUIdentifier adds a custom organizational unit identifier to the organization's MSP.`

File: configtx/application.go [hunk 5]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+func (m *ApplicationOrgMSP) AddOUIdentifier(ou membership.OUIdentifier) error {
+	msp, err := getMSPConfig(m.configGroup)
+	if err != nil {
+		return err
+	}
+
+	for _, o := range msp.OrganizationalUnitIdentifiers {
```
Comment (caod123) [inline configtx/application.go:571]: We may want to consider exactly what the behavior should be when adding duplicates for the name. A `membership.OUIdentifier` contains both a certificate as well as the string identifier. Effectively it can almost be treated as a map from identifier to the CA under which identities with that OU are validated, although for purposes of marshaling we treat it as a slice of a type that contains both the certificate and identifier. You may want to sync with MSP experts on whether an MSP can be configured with multiple certificates for the same OUIdentifier string. If that's not supported we may want to add an error here or alternatively change the functionality here to `SetOUIdentifier` where if the identifier string in the `membership.OUIdentifier` is already found in the existing config, it would just replace the certificate with the new one passed in since otherwise I see no path for modifying the existing OU identifiers without removing the old one to add a new one with the same string (effectively a replace).

File: configtx/application.go [hunk 6]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+			return nil
+		}
+	}
+
+	msp.OrganizationalUnitIdentifiers = append(msp.OrganizationalUnitIdentifiers, ou)
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveOUIdentifier removes a fabric organizational unit for the organization MSP.
```
Comment (caod123) [inline configtx/application.go:582]: `RemoveOUIdentifier removes an existing organizational unit identifier from the organization MSP.`

File: configtx/application.go [hunk 7]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+	if err != nil {
+		return err
+	}
+
+	msp.CryptoConfig = cryptoConfig
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddTLSRootCert adds an TLS root certificate trusted by the organization MSP.
```
Comment (caod123) [inline configtx/application.go:614]: `adds a TLS root`

File: configtx/application.go [hunk 8]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// RemoveTLSRootCert removes an trusted TLS root certificate from the organization MSP.
```
Comment (caod123) [inline configtx/application.go:637]: `removes a trusted`

File: configtx/application.go [hunk 9]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+		return err
+	}
+
+	certs := msp.TLSRootCerts[:]
+	for i, c := range msp.TLSRootCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
```
Comment (caod123) [inline configtx/application.go:650]: I checked the `msp.validateCACerts` function and it doesn't look like we validate TLS intermediate certs against the TLS root certs. @wlahti was this an oversight when we originally added the check for regular `RootCerts`/`IntermediateCerts`? https://github.com/hyperledger/fabric-config/blob/master/configtx/msp.go#L481-L517

Comment (caod123) [reply configtx/application.go:650]: We may want to still call `validateCACerts` here like my comment above for `RemoveRootCert` and in a followup story fix the above if `TLSIntermediateCerts` should also be validated against the `TLSRootCerts`

File: configtx/application.go [hunk 10]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+
+	msp.TLSRootCerts = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddTLSIntermediateCert adds an TLS intermediate cert trusted by the organization MSP.
```
Comment (caod123) [inline configtx/application.go:657]: `adds a TLS intermediate`

File: configtx/application.go [hunk 11]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+		return err
+	}
+
+	msp.NodeOUs.OrdererOUIdentifier = ordererOU
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// SetEnableNodeOUs sets the NodeOUs recognition, if NodeOUs recognition is enabled then an msp identity
+// that does not contain any of the specified OUs will be considered invalid.
```
Comment (caod123) [inline configtx/application.go:749]: `that does not contain exactly one of the fabric Node OU identifiers will be considered invalid.`

File: configtx/application.go [hunk 12]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+	if err != nil {
+		return err
+	}
+
+	msp.NodeOUs.Enable = isEnabled
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddCRL adds an CRL to the identity revocation list for the organization MSP.
```
Comment (caod123) [inline configtx/application.go:761]: `adds a CRL`

File: configtx/application.go [hunk 13]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+	if err != nil {
+		return err
+	}
+
+	msp.RevocationList = append(msp.RevocationList, crl)
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddCRLFromSigningIdentity creates an crl from the provided signing identity and certs then adds the CRL to the identity revocation list for the organization MSP.
```
Comment (caod123) [inline configtx/application.go:773]: `AddCRLFromSigningIdentity creates a CRL from the provided signing identity and associated certs and then adds the CRL to the identity revocation list for the organization MSP.`

File: configtx/msp.go [hunk 14]
```diff
@@ -139,6 +145,20 @@ func (m *MSP) isCACert(signingCert *x509.Certificate) error {
 	return fmt.Errorf("signing cert is not a root/intermediate cert for this MSP: %s", m.Name)
 }
 
+func (m *MSP) setMSPConfig(configGroup *cb.ConfigGroup) error {
+	mspConfig, err := newMSPConfig(*m)
```
Comment (caod123) [inline configtx/msp.go:149]: why is this needed?

File: configtx/msp.go [hunk 15]
```diff
@@ -139,6 +145,20 @@ func (m *MSP) isCACert(signingCert *x509.Certificate) error {
...
+	if err != nil {
+		return fmt.Errorf("new msp config: %v", err)
+	}
+
+	err = setValue(configGroup, mspValue(mspConfig), AdminsPolicyKey)
```
Comment (caod123) [inline configtx/msp.go:154]: why won't this work with just `mspValue(m)`?

File: configtx/msp.go [hunk 16]
```diff
@@ -139,6 +145,20 @@ func (m *MSP) isCACert(signingCert *x509.Certificate) error {
 	return fmt.Errorf("signing cert is not a root/intermediate cert for this MSP: %s", m.Name)
 }
 
+func (m *MSP) setMSPConfig(configGroup *cb.ConfigGroup) error {
```
Comment (caod123) [inline configtx/msp.go:148]: nit: This function naming is a bit odd. You would expect a `setMSPConfig` function to take an `mspConfig` in as a parameter. I'm fine leaving it as is but we may want to consider renaming this since it can get twisted/confusing.

File: configtx/msp.go (hunk 15 above)
Comment (stephyee) [reply configtx/msp.go:154]: No, `mspValue()` takes a proto.

File: configtx/msp.go (hunk 14 above)
Comment (stephyee) [reply configtx/msp.go:149]: To convert the MSP type to the MSPConfig proto type.

File: configtx/application.go [hunk 17]
```diff
@@ -39,6 +41,18 @@ type ApplicationOrg struct {
 	name     string
 }
 
+// ApplicationOrgMSP encapsulates the configuration functions used to modify an orderer organization's MSP.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go [hunk 18]
```diff
@@ -39,6 +41,18 @@ type ApplicationOrg struct {
...
+type ApplicationOrgMSP struct {
+	configGroup *cb.ConfigGroup
+}
+
+// MSP returns an ApplicationOrgMSP that can be used to configure an MSP.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go [hunk 19]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
 	return nil
 }
 
-// CreateMSPCRL creates a CRL that revokes the provided certificates
-// for the specified application org signed by the provided SigningIdentity.
-func (a *ApplicationOrg) CreateMSPCRL(signingIdentity *SigningIdentity, certs ...*x509.Certificate) (*pkix.CertificateList, error) {
-	msp, err := a.MSP()
+// Configuration returns the MSP value for an orderer organization in the
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go [hunk 20]
```diff
@@ -423,15 +431,359 @@ func (a *ApplicationOrg) setMSPConfig(updatedMSP MSP) error {
...
+			break
+		}
+	}
+
+	msp.Admins = certs
+
+	return msp.setMSPConfig(m.configGroup)
+}
+
+// AddRootCert adds an root certificate trusted by the organization MSP.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 1 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 3 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 4 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 6 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 7 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 8 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 10 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 11 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 12 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 13 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/application.go (hunk 5 above)
Comment (stephyee) [reply configtx/application.go]: Are you okay with investigating this separately?

File: configtx/application.go (hunk 9 above)
Comment (stephyee) [reply configtx/application.go]: added `validateCACerts()`

File: configtx/application.go (hunk 2 above)
Comment (stephyee) [reply configtx/application.go]: added `validateCACerts()`

File: configtx/application.go [hunk 21]
```diff
@@ -39,6 +37,18 @@ type ApplicationOrg struct {
 	name     string
 }
 
+// ApplicationOrgMSP encapsulates the configuration functions used to modify an application organization's MSP.
```
Comment (caod123) [inline configtx/application.go:40]: This type is unused now and can be deleted.

File: configtx/orderer.go [hunk 22]
```diff
@@ -62,6 +61,18 @@ type OrdererOrg struct {
 	name     string
 }
 
+// OrdererOrgMSP encapsulates the configuration functions used to modify an orderer organization's MSP.
+type OrdererOrgMSP struct {
```
Comment (caod123) [inline configtx/orderer.go:65]: This type is unused now and can be deleted.

File: configtx/msp.go [hunk 23]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
 const YEAR = 365 * 24 * time.Hour
 
+// OrganizationMSP encapsulates the configuration functions used to modify an organization MSP.
+type OrganizationMSP struct {
+	configGroup *cb.ConfigGroup
+}
+
+// CreateMSPCRL creates a CRL that revokes the provided certificates
+// for the specified organization's msp signed by the provided SigningIdentity.
+func (m *MSP) CreateMSPCRL(signingIdentity *SigningIdentity, certs ...*x509.Certificate) (*pkix.CertificateList, error) {
```
Comment (caod123) [inline configtx/msp.go:85]: nit: move this down to where the other `MSP` functions are grouped

File: configtx/msp.go [hunk 24]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+	return m.newMSPCRL(signingIdentity, certs...)
+}
+
+// Configuration returns the MSP value for an orderer organization in the
```
Comment (caod123) [inline configtx/msp.go:89]: `Configuration returns the MSP value for an organization in the updated config.`

File: configtx/msp.go [hunk 25]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+
+	err = msp.validateCACerts()
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// RemoveIntermediateCert removes a trusted intermediate certificate trusted by the organization MSP.
```
Comment (caod123) [inline configtx/msp.go:204]: lot of `trusted` here :) maybe remove one of them.
`RemoveIntermediateCert removes an intermediate certificate trusted by the organization MSP.` or `RemoveIntermediateCert removes a trusted intermediate certificate from the organization MSP.` Your choice, same comment applies to the other comments that use `trusted` redundantly.

File: configtx/msp.go [hunk 26]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+	for i, c := range msp.IntermediateCerts {
+		if c.Equal(cert) {
+			certs = append(certs[:i], certs[i+1:]...)
+			break
+		}
+	}
+
+	msp.IntermediateCerts = certs
+
+	err = msp.validateCACerts()
```
Comment (caod123) [inline configtx/msp.go:221]: This probably isn't necessary but also doesn't hurt to have whenever we touch the ca certs/intermediate cas.

File: configtx/msp.go [hunk 27]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+	if err != nil {
+		return err
+	}
+
+	return msp.setConfig(m.configGroup)
+}
+
+// AddOUIdentifier adds a custom organizational unit identifier to the organization's MSP.
```
Comment (caod123) [inline configtx/msp.go:229]: looks like we use `organization MSP` instead of `organization's MSP` in the other comments. Let's stick with `organization MSP` here as well.

File: configtx/msp.go [hunk 28]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+		return err
+	}
+
+	for _, o := range msp.OrganizationalUnitIdentifiers {
+		if reflect.DeepEqual(o, ou) {
+			return nil
+		}
+	}
+
+	msp.OrganizationalUnitIdentifiers = append(msp.OrganizationalUnitIdentifiers, ou)
```
Comment (caod123) [inline configtx/msp.go:242]: I'm still unsure how duplicate identifiers should be treated (see my comment from the previous review).

File: configtx/msp.go [hunk 29]
```diff
@@ -74,6 +75,393 @@ type MSP struct {
...
+
+	for _, c := range msp.TLSIntermediateCerts {
+		if c.Equal(cert) {
+			return nil
+		}
+	}
+
+	msp.TLSIntermediateCerts = append(msp.TLSIntermediateCerts, cert)
+
+	err = msp.validateCACerts()
```
Comment (caod123) [inline configtx/msp.go:342]: Can you add a followup story for adding TLS intermediate cert validation up the chain to the `validateCACerts`?

Comment (stephyee) [reply configtx/msp.go:342]: https://jira.hyperledger.org/browse/FAB-18099 added to the backlog.

File: configtx/msp.go (hunk 28 above)
Comment (stephyee) [reply configtx/msp.go:242]: As I noted in my previous comment, let's follow up on this and defer validation in the library.

File: configtx/consortiums.go [hunk 30]
```diff
@@ -43,6 +41,18 @@ type ConsortiumOrg struct {
 	name     string
 }
 
+// ConsortiumOrgMSP encapsulates the configuration functions used to modify an consortium organization's MSP.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/consortiums.go]: 🛠️

File: configtx/msp.go (hunk 25 above)
Comment (stephyee) [reply configtx/msp.go]: 🛠️ looks like this was the only place with multiple uses of trusted

File: configtx/msp.go (hunk 26 above)
Comment (stephyee) [reply configtx/msp.go:221]: 🤔 I thought you pointed this out in a previous comment.

File: configtx/msp.go (hunk 27 above)
Comment (stephyee) [reply configtx/msp.go]: 🛠️

File: configtx/application.go (hunk 21 above)
Comment (stephyee) [reply configtx/application.go]: 🛠️

File: configtx/orderer.go (hunk 22 above)
Comment (stephyee) [reply configtx/orderer.go]: 🛠️

File: configtx/msp.go (hunk 24 above)
Comment (stephyee) [reply configtx/msp.go]: 🛠️

File: configtx/msp.go (hunk 23 above)
Comment (stephyee) [reply configtx/msp.go]: 🛠️

File: configtx/orderer.go [hunk 31]
```diff
@@ -661,39 +666,14 @@ func (o *OrdererOrg) SetMSP(updatedMSP MSP) error {
 		return err
 	}
 
-	err = o.setMSPConfig(updatedMSP)
+	err = updatedMSP.setConfig(o.orgGroup)
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/orderer.go:664]: 👌

File: configtx/orderer.go [hunk 32]
```diff
@@ -62,6 +61,18 @@ type OrdererOrg struct {
 	name     string
 }
 
+// OrdererOrgMSP encapsulates the configuration functions used to modify an orderer organization's MSP.
+type OrdererOrgMSP struct {
+	configGroup *cb.ConfigGroup
+}
+
+// MSP returns the MSP configuration for an orderer organization.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/orderer.go]: 🛠️

File: configtx/consortiums.go [hunk 33]
```diff
@@ -43,6 +41,18 @@ type ConsortiumOrg struct {
...
+type ConsortiumOrgMSP struct {
+	configGroup *cb.ConfigGroup
+}
+
+// MSP returns the MSP configuration for an consortium organization.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/consortiums.go:44]: 🛠️

File: configtx/application.go [hunk 34]
```diff
@@ -39,6 +37,18 @@ type ApplicationOrg struct {
...
+type ApplicationOrgMSP struct {
+	configGroup *cb.ConfigGroup
+}
+
+// MSP returns the MSP configuration for an application organization.
```
[the remark this answers is not in the document]
Comment (stephyee) [reply configtx/application.go]: 🛠️

**Comment by caod123** [comment]
As suggested in the conversation above, lets merge this for now and refactor the `toProto` logic later.

File: configtx/application.go (hunk 9 above)
Comment (wlahti) [reply configtx/application.go]: > I checked the `msp.validateCACerts` function and it doesn't look like we validate TLS intermediate certs against the TLS root certs. @wlahti was this an oversight when we originally added the check for regular `RootCerts`/`IntermediateCerts`? https://github.com/hyperledger/fabric-config/blob/master/configtx/msp.go#L481-L517

Yes, it looks like this was an oversight, likely due to me not being very specific in the JIRA task and specifying we should validate the root/intermediate CA certs for both regular and TLS CA certificates. Good catch. :)
